### ISSUE TITLE
Speed and Efficiency of Depthwise separable operation?

### ISSUE THREAD

**Issue by kennyseb** [shrubb/box-convolutions#18]
As far as modern libraries are concerned, there is not much support for depth-wise separable operations, i.e. we cannot write custom operations that can be done depthwise. Only convolutions are supported.

How did you apply M box convolutions to each of the N input filters, to generate NM output filters? 
How is the different than using a for loop over the N input filters, applying M box convs on each one, and concatenating all the results?

**Comment by shrubb** [comment]
Sorry, couldn't really understand your point and question. "Only convolutions are supported" is wrong: there are countless depthwise operations, for example
* all elementwise operations,
* max pooling, average pooling,
* `nn.Conv2d(n, m, k, groups=n)`,
* batch normalization etc.
which means we of course *can* create other depthwise operations, and it is standard to do so. :)

Not sure what you mean by "depthwise separable operations". You might be confusing depthwise convolutions and depthwise separable convolutions (which is a sequence of layers that is overall not depthwise).

As for "how is it different than using a for loop": you are right, it's exactly how it works, there's no difference.

**Comment by kennyseb** [comment]
Thanks, this makes it clear.
One more question, consider the input layer to the BoxConv Layer is of dimension NxRxC, where n=num of channels, r & c = spatial dimensions.

How do you regress the variables corresponding to each box?
Did you use a localisation network like in a STN, e.g. [here](https://pytorch.org/tutorials/intermediate/spatial_transformer_tutorial.html) which takes all the NxRxC neurons as inputs and outputs the corresponding 4xP values ( P=output boxes).

Can you please give an idea of the architecture of the BoxConv module? I realise that it has not been provided in the paper.

Thanks a lot!

**Comment by shrubb** [comment]
> How do you regress the variables corresponding to each box?

We don't regress them. They are `BoxConv2d`'s 's parameters (like `nn.Linear.weight` or `nn.Conv2d.weight`), which are initialized randomly, stored in the module and learned via gradient descent.

> Can you please give an idea of the architecture of the BoxConv module? I realise that it has not been provided in the paper.

The module's architecture is the main point of our work, so it is actually described thoroughly in the paper and is implemented in this repository (so you can read the code to learn what's going on).

**Comment by kennyseb** [comment]
I see, does that mean that the box parameters (x, y, w, h) are learned throughout the training, and remain constant while testing? i.e., the generated boxes are same for all the images, is that right?

**Comment by shrubb** [comment]
Exactly.
